packages/backend/server/src/models/copilot-workspace.ts

```diff
@@ -0,0 +1,209 @@
+  ) {
+    const groups = embeddings.map(e =>
+      [
+        workspaceId,
+        fileId,
+        e.index,
+        e.content,
+        Prisma.raw(`'[${e.embedding.join(',')}]'`),
+      ].filter(v => v !== undefined)
+    );
+    return Prisma.join(groups.map(row => Prisma.sql`(${Prisma.join(row)})`));
+  }
+
+  private async getPayload(workspaceId: string) {
+    const payload = await this.db.aiWorkspaceConfig
+      .findFirst({
+        where: {
+          workspaceId,
+        },
+        select: {
+          payload: true,
+        },
+      })
+      .then(r => r?.payload ?? { files: {} });
+
+    return CopilotWorkspaceConfigPayloadSchema.parse(payload);
+  }
+
+  @Transactional()
+  async addWorkspaceFile(
+    workspaceId: string,
+    file: CopilotWorkspaceFile,
+    embeddings: Embedding[]
+  ) {
+    await this.init(workspaceId);
+
+    const { fileId, ...fileData } = file;
+    const payload = await this.getPayload(workspaceId);
+    if (!payload.files) {
+      payload.files = {};
+    }
+    if (payload.files[fileId]) {
+      return;
+    }
+    payload.files[fileId] = fileData;
+
+    await this.db.aiWorkspaceConfig.update({
+      where: {
+        workspaceId,
+      },
+      data: {
+        payload,
+        updatedAt: new Date(),
+      },
+    });
+
+    const values = this.processEmbeddings(workspaceId, fileId, embeddings);
+    await this.db.$executeRaw`
+        INSERT INTO "ai_workspace_file_embeddings"
+        ("workspace_id", "file_id", "chunk", "content", "embedding") VALUES ${values}
+        ON CONFLICT (workspace_id, file_id, chunk) DO NOTHING;
+    `;
+  }
+
+  async listWorkspaceFiles(
+    workspaceId: string
+  ): Promise<CopilotWorkspaceFile[]> {
+    const row = await this.getPayload(workspaceId);
+    return Object.entries(row.files || {}).map(([fileId, file]) => ({
+      fileId,
+      ...file,
+    }));
+  }
+
+  async matchWorkspaceFileEmbedding(
+    workspaceId: string,
+    embedding: number[],
+    topK: number,
+    threshold: number
+  ): Promise<FileChunkSimilarity[]> {
+    const similarityChunks = await this.db.$queryRaw<
+      Array<FileChunkSimilarity>
+    >`
+      SELECT "file_id" as "fileId", "chunk", "content", "embedding" <=> ${embedding}::vector as "distance" 
+      FROM "ai_workspace_file_embeddings"
+      WHERE workspace_id = ${workspaceId}
+      ORDER BY "distance" ASC
+      LIMIT ${topK};
+    `;
+    return similarityChunks.filter(c => Number(c.distance) <= threshold);
+  }
+
+  async removeWorkspaceFile(workspaceId: string, fileId: string) {
+    const payload = await this.getPayload(workspaceId);
+    if (!payload.files) {
+      payload.files = {};
+    }
+    if (!payload.files[fileId]) {
+      return;
+    }
+    delete payload.files[fileId];
+
+    await this.db.aiWorkspaceConfig.update({
+      where: {
+        workspaceId,
+      },
+      data: {
+        payload,
+        updatedAt: new Date(),
+      },
+    });
+
+    await this.db.$executeRaw`
```

Comment:
Do all table queries involving `vector` type field require handwritten SQL?

Reply:
yes, prisma not have vector support yet